### DISCUSSION TITLE
boot.wim (WinPE 10.0.19041.1 32-bit) BSOD

### DISCUSSION THREAD

**Discussion by sm1000** [ipxe/wimboot#13] [General]
winboot (v.2.6.0) commit [d9158db](https://github.com/ipxe/wimboot/commit/d9158db8cbe9e6d56f1ae94e62f170f5d3998ccb#diff-e02e2bfc8776198b534bcd0e6501bc2c2264de284def8c6fa960108580ca36f9) works fine when booting WinPE 10.0.19041.1 **32-bit**.

winboot (v.2.7.0 and v.2.7.1) commits [699f810](https://github.com/ipxe/wimboot/commit/699f81059fe879b44e8de3a4e1a12dbd0c88d01d#diff-e02e2bfc8776198b534bcd0e6501bc2c2264de284def8c6fa960108580ca36f9) and [60d991b](https://github.com/ipxe/wimboot/commit/60d991bd6830c092a4a406dd1b3421817c3b564a#diff-e02e2bfc8776198b534bcd0e6501bc2c2264de284def8c6fa960108580ca36f9) are crashing the WinPE 10.0.19041.1 **32-bit** with a BSOD. Windows error code 0xc000000d.

Here are the links to the latest WinPE version:
[Windows ADK for Windows 10](https://go.microsoft.com/fwlink/?linkid=2120254)
[Windows PE add-on for the ADK](https://go.microsoft.com/fwlink/?linkid=2120253)

NOTE: WinPE 10.0.19041.1 **64-bit** works fine on all winboot releases.

My ipxe script is:
**kernel ${file-location}/wimboot-271**
**initrd ${file-location}/Boot/bcd          bcd**
**initrd ${file-location}/Boot/boot.sdi     boot.sdi**
**initrd ${file-location}/Sources/boot.wim  boot.wim**
**boot**

**Comment by mcb30** [comment]
The automated test runs include a boot of the 32-bit Windows 10.0.19041.1 `boot.wim` as found on the installation ISO image.  You can see the test pass from the v2.7.1 commit 60d991b at https://github.com/ipxe/wimboot/runs/1880022181 - test logs and screenshots are included in the [`qa-results`](https://github.com/ipxe/wimboot/suites/2019183498/artifacts/40481314) artifact.

Could you check that you are able to successfully boot the original `boot.wim` as found on the installation ISO?

If the installation ISO works but your own ADK-created image continues to fail, then please provide the commands required to recreate your images using the ADK.

**Comment by sm1000** [comment]
The original [32-bit installation WIM](https://www.microsoft.com/en-us/software-download/windows10) ("Sources\install.wim\[index]\Windows\System32\Recovery\Winre.wim") has the same problem on my virtual machine and on my physical machine (Dell Optiplex 3070).

All wimboot releases are from https://github.com/ipxe/wimboot (I did not build them).

The ipxe.pxe (commit [900f1f9](https://github.com/ipxe/ipxe/commit/900f1f98d3dbb64803e427318ad61c9d7aa6f9bb#diff-25a6634263c1b1f6fc4697a04e2b9904ea4b042a89af59dc93ec1f5d44848a26)) is build on [Debian 10.8](https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/debian-10.8.0-amd64-netinst.iso) as "bin-i386-pcbios/ipxe.pxe"
with the following configuration ("/ipxe/src/config/local"):
/* console */
#define	CONSOLE_PCBIOS
#define	CONSOLE_FRAMEBUFFER
#define	KEYBOARD_MAP	us
#define	LOG_LEVEL	LOG_NONE
/* crypto */
#define TLS_VERSION_MIN TLS_VERSION_TLS_1_1
#define CRYPTO_PUBKEY_RSA
#define CRYPTO_CIPHER_AES_CBC
#define CRYPTO_DIGEST_SHA1
#define CRYPTO_DIGEST_SHA224
#define CRYPTO_DIGEST_SHA256
#define CRYPTO_DIGEST_SHA384
#define CRYPTO_DIGEST_SHA512
#define TIMESTAMP_ERROR_MARGIN ( ( 12 * 60 + 30 ) * 60 )
#define CROSSCERT "http://ca.ipxe.org/auto"
#define OCSP_CHECK
/* general */
#define BANNER_TIMEOUT		20
#define ROM_BANNER_TIMEOUT	( 2 * BANNER_TIMEOUT )
#define	NET_PROTO_IPV4
#define	NET_PROTO_IPV6
#undef	NET_PROTO_FCOE
#define	NET_PROTO_STP
#define	NET_PROTO_LACP
#define	NET_PROTO_EAPOL
#define	PXE_STACK
#define	PXE_MENU
#define	DOWNLOAD_PROTO_TFTP
#define	DOWNLOAD_PROTO_HTTP
#define	DOWNLOAD_PROTO_HTTPS
#define	DOWNLOAD_PROTO_FTP
#define	DOWNLOAD_PROTO_SLAM
#define	DOWNLOAD_PROTO_NFS
#define	DOWNLOAD_PROTO_FILE
#define	SANBOOT_PROTO_ISCSI
#define	SANBOOT_PROTO_HTTP
#define HTTP_AUTH_BASIC
#define HTTP_AUTH_DIGEST
#define HTTP_AUTH_NTLM
#define HTTP_ENC_PEERDIST
#define HTTP_HACK_GCE
#define	CRYPTO_80211_WEP
#define	CRYPTO_80211_WPA
#define	CRYPTO_80211_WPA2
#define	DNS_RESOLVER
#define	IMAGE_MULTIBOOT
#define	IMAGE_PXE
#define	IMAGE_SCRIPT
#define	IMAGE_BZIMAGE
#define	IMAGE_COMBOOT
#define	IMAGE_PNM
#define	IMAGE_PNG
#define	IMAGE_DER
#define	IMAGE_PEM
#define	AUTOBOOT_CMD
#define	NVO_CMD
#define	CONFIG_CMD
#define	IFMGMT_CMD
#define	IWMGMT_CMD
#define IBMGMT_CMD
#define FCMGMT_CMD
#define	ROUTE_CMD
#define IMAGE_CMD
#define DHCP_CMD
#define SANBOOT_CMD
#define MENU_CMD
#define LOGIN_CMD
#define SYNC_CMD
#define SHELL_CMD
#define NSLOOKUP_CMD
#define TIME_CMD
#define DIGEST_CMD
#define LOTEST_CMD
#define VLAN_CMD
#define PXE_CMD
#define REBOOT_CMD
#define POWEROFF_CMD
#define IMAGE_TRUST_CMD
#define PCI_CMD
#define PARAM_CMD
#define NEIGHBOUR_CMD
#define PING_CMD
#define CONSOLE_CMD
#define IPSTAT_CMD
#define PROFSTAT_CMD
#define NTP_CMD
#define CERT_CMD
#define IMAGE_MEM_CMD
#undef	NONPNP_HOOK_INT19
#define	AUTOBOOT_ROM_FILTER
#define VNIC_IPOIB
#undef	ERRMSG_80211
#undef	BUILD_SERIAL
#undef	BUILD_ID
#undef	NULL_TRAP
#undef	GDBSERIAL
#undef	GDBUDP
#define	TIVOLI_VMM_WORKAROUND
/* settings */
#define	PCI_SETTINGS
#define	CPUID_SETTINGS
#define	VMWARE_SETTINGS
#define	VRAM_SETTINGS
#define	ACPI_SETTINGS
/* usb */
#define	USB_HCD_XHCI
#define	USB_HCD_EHCI
#define	USB_HCD_UHCI
#define	USB_KEYBOARD
#define	USB_EFI


Let me know if you need more details from my test environment.

**Reply by mcb30** [reply]
I have verified that the following works as expected:

- Extract `sources/install.wim` from the installation ISO
- Extract `Windows/System32/Recovery/Winre.wim` from `sources/install.wim`
- Boot using the iPXE script:
  ```
  #!ipxe
  kernel wimboot
  initrd -n boot.wim Winre.wim boot.wim
  boot
  ```

This boots successfully into WinRE, using the `BCD` and `boot.sdi` files present inside the `WinRE.wim` file itself.  (Note that the embedded `BCD` file refers to `boot.wim` rather than `Winre.wim`, hence the name changes in the `initrd` line.)

I note that your original example includes a custom `BCD` file of unknown origin.  Try deleting these two lines from your iPXE script:
```
initrd ${file-location}/Boot/bcd bcd
initrd ${file-location}/Boot/boot.sdi boot.sdi
```

**Comment by sm1000** [comment]
Hi Michael,

This is exactly what I was doing.
The two extra files (**bcd** and **boot.sdi**) are extracted by me from the **Winre.wim** while looking for a solution on the problem.

I applied your script and the result is the same - crash with Windows error code 0xc000000d.

I am available if you need any other information from my system.

NOTE: My **Winre.wim** is taken from index # 6.

Regards.

**Comment by sm1000** [comment]
Based on the time when the Windows crash error appears, I would assume that it is at the moment when Windows tries to switch into the graphic mode...

I could be wrong...

**Comment by mcb30** [comment]
I have been able to reproduce the issue.

For the record: the use of the ADK and the extraction of `Winre.wim` from the installation ISO `install.wim` are irrelevant.  The error can be reproduced using the `boot.wim` image as found on the installation ISO, which is why I did ask you to please test that in my very first reply:

> Could you check that you are able to successfully boot the original `boot.wim` as found on the installation ISO?

**Reply by mcb30** [reply]
The following `.yml` file defines a test case that can reproduce the problem:

```yaml
name: Windows 10 (High RAM, 32-bit)
version: win10
arch: x86
memory: 5120
logcheck:
  - 'Found initrd at \[0x[0-9a-f]{8},0x[8-9a-f][0-9a-f]{7}\)'
  - 'Placing initrd at \[0x[0-9a-f]{8},0x[0-7][0-9a-f]{7}\)'
  - 'Placing initrd at physical \[0x[1-9a-f][0-9a-f]{8,},'
```

This test passes for Windows 7 but fails for Windows 8.1 or Windows 10 with the `c000000d` error.

**Reply by sm1000** [reply]
> I have been able to reproduce the issue.
> 
> For the record: the use of the ADK and the extraction of `Winre.wim` from the installation ISO `install.wim` are irrelevant. The error can be reproduced using the `boot.wim` image as found on the installation ISO, which is why I did ask you to please test that in my very first reply:
> 
> > Could you check that you are able to successfully boot the original `boot.wim` as found on the installation ISO?

This is a great news.

For the record: The boot.wim file inside the installation ISO is actually a dual-boot WIM with an index in the main folder (**[1].xml**). Please let me know how I can use it to boot from it so I can do it next time.

Thank you.

**Reply by mcb30** [reply]
> For the record: The boot.wim file inside the installation ISO is actually a dual-boot WIM with an index in the main folder (**[1].xml**). Please let me know how I can use it to boot from it so I can do it next time.

Assuming that you have the installation ISO unpacked into the current directory alongside the `wimboot` binary, then you can just use the script
```
#!ipxe
kernel wimboot
initrd sources/boot.wim boot.wim
boot
```

The fact that it contains multple images is not a problem: the `.wim` format includes a "boot index" field that will be used automatically to select the image to be booted.  (In case it's useful in future: you can also override the boot index via the `index=N` parameter as documented at https://ipxe.org/wimboot#multi-image_wim_files)

**Reply by sm1000** [reply]
Thank you Michael.

**Comment by mcb30** [comment]
Raised as issue #15 since this is a defect

**Reply by mcb30** [reply]
Should be fixed as of commit https://github.com/ipxe/wimboot/commit/c36ca86
